File: Linked_List_free_code_camp/1.linked_list_traversal.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;

    Node(int val, Node* next = NULL) {
        this->val = val;
        this->next = next;
    }
};

void printLinkedListIterative(Node* head) {
    Node* t = head;
    while (t) {
        cout << t->val << " ";
        t = t->next;
    }
}

void printLinkedListRecursive(Node* head) {
    if (head == NULL) return;

    cout << head->val << " ";
    printLinkedListRecursive(head->next);
}

int main() {

    Node* head =
        new Node(10,
            new Node(20,
                new Node(30,
                    new Node(40))));

    printLinkedListIterative(head);
    cout << endl;
    printLinkedListRecursive(head);


    return 0;
}

File: Linked_List_free_code_camp/1.linked_list_traversal.js
class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

const head =
    new Node('A',
        new Node('B',
            new Node('C',
                new Node('D', null))));


// A -> B -> C -> D -> null

const printLinkedListIterative = (head) => {
    let curr = head;
    while (curr) {
        console.log(curr.val);
        curr = curr.next;
    }
};

// printLinkedListIterative(head);


const printLinkedListRecursive = (head) => {
    if (head == null) return;
    console.log(head.val);
    printLinkedListRecursive(head.next);
};

printLinkedListRecursive(head);
